File: myfschoolse1915/lib/vn/edu/fpt/controller/club_controller.dart
import 'package:myfschoolse1915/vn/edu/fpt/utils/api_config.dart';
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:myfschoolse1915/vn/edu/fpt/model/club_model.dart';

class ClubController {
  final String baseUrl = '${ApiConfig.baseUrl}/api';

  Future<List<ClubModel>> fetchClubs() async {
    try {
      final response = await http.get(Uri.parse('$baseUrl/club'));
      if (response.statusCode == 200) {
        List<dynamic> data = jsonDecode(response.body);
        return data.map((json) => ClubModel.fromJson(json)).toList();
      } else {
        throw Exception('Failed to load clubs');
      }
    } catch (e) {
      print('Error fetching clubs: $e');
      throw e;
    }
  }

  Future<ClubStatsModel> fetchClubStats() async {
    try {
      final response = await http.get(Uri.parse('$baseUrl/club/stats'));
      if (response.statusCode == 200) {
        return ClubStatsModel.fromJson(jsonDecode(response.body));
      } else {
        throw Exception('Failed to load club stats');
      }
    } catch (e) {
      print('Error fetching club stats: $e');
      throw e;
    }
  }

  Future<bool> createClub(String name, String type, String status, String description, int? presidentId) async {
    final response = await http.post(
      Uri.parse('$baseUrl/club/add'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'name': name,
        'type': type,
        'status': status,
        'description': description,
        'presidentId': presidentId
      }),
    );
    return response.statusCode == 200;
  }

  Future<bool> updateClub(int id, String name, String type, String status, String description, int? presidentId) async {
    final response = await http.put(
      Uri.parse('$baseUrl/club/update/$id'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'name': name,
        'type': type,
        'status': status,
        'description': description,
        'presidentId': presidentId
      }),
    );
    return response.statusCode == 200;
  }

  Future<bool> deleteClub(int id) async {
    final response = await http.delete(Uri.parse('$baseUrl/club/delete/$id'));
    return response.statusCode == 200;
  }

  Future<List<dynamic>> fetchClubMembers(int id) async {
    final response = await http.get(Uri.parse('$baseUrl/club/$id/members'));
    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    }
    throw Exception('Failed to load members');
  }

  Future<bool> addClubMember(int clubId, int userId) async {
    final response = await http.post(
      Uri.parse('$baseUrl/club/$clubId/add-member'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'userId': userId}),
    );
    return response.statusCode == 200;
  }

  Future<bool> removeClubMember(int clubId, int userId) async {
    final response = await http.delete(Uri.parse('$baseUrl/club/$clubId/remove-member/$userId'));
    return response.statusCode == 200;
  }

  Future<bool> joinRequest(int clubId, int userId, String reason) async {
    final response = await http.post(
      Uri.parse('$baseUrl/club/$clubId/join-request'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'userId': userId, 'reason': reason}),
    );
    if (response.statusCode != 200) {
      final msg = jsonDecode(response.body)['message'];
      throw Exception(msg ?? 'Gửi yêu cầu thất bại');
    }
    return true;
  }

  Future<List<dynamic>> fetchPendingRequests(int clubId) async {
    final response = await http.get(Uri.parse('$baseUrl/club/$clubId/pending-requests'));
    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    }
    throw Exception('Failed to load requests');
  }

  Future<bool> transferPresident(int clubId, int currentId, int newId) async {
    final response = await http.post(
      Uri.parse('$baseUrl/club/$clubId/transfer-president'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'currentPresidentId': currentId, 'newPresidentId': newId}),
    );
    if (response.statusCode != 200) {
      final msg = jsonDecode(response.body)['message'];
      throw Exception(msg ?? 'Chuyển quyền thất bại');
    }
    return true;
  }

  Future<bool> approveRequest(int requestId) async {
    final response = await http.post(Uri.parse('$baseUrl/AppRequest/$requestId/approve'));
    if (response.statusCode != 200) {
      throw Exception('Duyệt đơn thất bại');
    }
    return true;
  }

  Future<bool> rejectRequest(int requestId) async {
    final response = await http.post(Uri.parse('$baseUrl/AppRequest/$requestId/reject'));
    if (response.statusCode != 200) {
      throw Exception('Từ chối đơn thất bại');
    }
    return true;
  }
}

File: backend/MyBackendAPI/Controllers/ClubController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using MyBackendAPI.Data;
using MyBackendAPI.Models;

namespace MyBackendAPI.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class ClubController : ControllerBase
    {
        private readonly AppDbContext _context;

        public ClubController(AppDbContext context)
        {
            _context = context;
        }

        // GET: api/club
        [HttpGet]
        public async Task<IActionResult> GetClubs()
        {
            var clubs = await _context.Clubs
                .Include(c => c.President)
                .Include(c => c.Members)
                .ToListAsync();

            var result = clubs.Select(c => new
            {
                c.Id,
                c.Name,
                c.Code,
                c.Type,
                c.Status,
                c.Description,
                PresidentId = c.PresidentId,
                PresidentName = c.President?.FullName,
                MemberCount = c.Members.Count,
                MemberIds = c.Members.Select(m => m.UserId).ToList()
            });

            return Ok(result);
        }

        // GET: api/club/stats
        [HttpGet("stats")]
        public async Task<IActionResult> GetClubStats()
        {
            var totalClubs = await _context.Clubs.CountAsync();
            var totalMembers = await _context.ClubMembers.Select(cm => cm.UserId).Distinct().CountAsync();
            
            // Hardcoded for now based on mockup
            var eventsThisMonth = 18;
            var pendingProposals = 5;

            return Ok(new
            {
                totalClubs,
                totalMembers,
                eventsThisMonth,
                pendingProposals
            });
        }

        // POST: api/club/add
        [HttpPost("add")]
        public async Task<IActionResult> CreateClub([FromBody] ClubCreateDto dto)
        {
            var club = new Club
            {
                Name = dto.Name,
                Type = dto.Type,
                Status = dto.Status,
                Description = dto.Description,
                PresidentId = dto.PresidentId
            };

            // Auto-generate Code: CLB + number (e.g., CLB01)
            var count = await _context.Clubs.CountAsync();
            club.Code = $"CLB{(count + 1).ToString("D2")}";

            _context.Clubs.Add(club);
            await _context.SaveChangesAsync();

            return Ok(new { message = "Thêm câu lạc bộ thành công", id = club.Id });
        }

        // PUT: api/club/update/{id}
        [HttpPut("update/{id}")]
        public async Task<IActionResult> UpdateClub(int id, [FromBody] ClubUpdateDto dto)
        {
            var club = await _context.Clubs.FindAsync(id);
            if (club == null) return NotFound(new { message = "Không tìm thấy câu lạc bộ" });

            club.Name = dto.Name;
            club.Type = dto.Type;
            club.Status = dto.Status;
            club.Description = dto.Description;
            club.PresidentId = dto.PresidentId;

            await _context.SaveChangesAsync();

            return Ok(new { message = "Cập nhật thành công" });
        }

        // DELETE: api/club/delete/{id}
        [HttpDelete("delete/{id}")]
        public async Task<IActionResult> DeleteClub(int id)
        {
            var club = await _context.Clubs.FindAsync(id);
            if (club == null) return NotFound(new { message = "Không tìm thấy câu lạc bộ" });

            _context.Clubs.Remove(club);
            await _context.SaveChangesAsync();

            return Ok(new { message = "Xóa thành công" });
        }

        // GET: api/club/{id}/members
        [HttpGet("{id}/members")]
        public async Task<IActionResult> GetClubMembers(int id)
        {
            var members = await _context.ClubMembers
                .Include(cm => cm.User)
                .Where(cm => cm.ClubId == id)
                .Select(cm => new {
                    cm.UserId,
                    Name = cm.User.FullName,
                    Email = cm.User.Email,
                    cm.JoinedDate
                })
                .ToListAsync();

            return Ok(members);
        }

        // POST: api/club/{id}/add-member
        [HttpPost("{id}/add-member")]
        public async Task<IActionResult> AddMember(int id, [FromBody] AddMemberDto dto)
        {
            if (!await _context.Clubs.AnyAsync(c => c.Id == id))
                return NotFound(new { message = "Không tìm thấy câu lạc bộ" });

            if (!await _context.Users.AnyAsync(u => u.Id == dto.UserId))
                return NotFound(new { message = "Không tìm thấy người dùng" });

            if (await _context.ClubMembers.AnyAsync(cm => cm.ClubId == id && cm.UserId == dto.UserId))
                return BadRequest(new { message = "Người dùng đã là thành viên của câu lạc bộ này" });

            var member = new ClubMember
            {
                ClubId = id,
                UserId = dto.UserId,
                JoinedDate = DateTime.UtcNow
            };

            _context.ClubMembers.Add(member);
            await _context.SaveChangesAsync();

            return Ok(new { message = "Thêm thành viên thành công" });
        }

        // DELETE: api/club/{id}/remove-member/{userId}
        [HttpDelete("{id}/remove-member/{userId}")]
        public async Task<IActionResult> RemoveMember(int id, int userId)
        {
            var member = await _context.ClubMembers
                .Include(cm => cm.Club)
                .FirstOrDefaultAsync(cm => cm.ClubId == id && cm.UserId == userId);

            if (member == null)
                return NotFound(new { message = "Thành viên không tồn tại trong câu lạc bộ" });

            var clubName = member.Club?.Name ?? "Câu lạc bộ";

            _context.ClubMembers.Remove(member);
            
            _context.Notifications.Add(new Notification
            {
                UserId = userId,
                Title = "Thông báo từ CLB",
                Message = $"Bạn đã bị xóa khỏi {clubName}.",
                CreatedAt = DateTime.UtcNow
            });

            await _context.SaveChangesAsync();

            return Ok(new { message = "Xóa thành viên thành công" });
        }

        // POST: api/club/{id}/join-request
        [HttpPost("{id}/join-request")]
        public async Task<IActionResult> JoinRequest(int id, [FromBody] JoinRequestDto dto)
        {
            var club = await _context.Clubs.FindAsync(id);
            if (club == null) return NotFound(new { message = "Không tìm thấy câu lạc bộ" });

            if (club.PresidentId == null) 
                return BadRequest(new { message = "Câu lạc bộ chưa có chủ nhiệm để duyệt đơn" });

            var existingRequest = await _context.AppRequests
                .FirstOrDefaultAsync(r => r.Category == "CLB_JOIN" && r.ReferenceId == id && r.SenderId == dto.UserId && r.Status == "Chờ duyệt");
            
            if (existingRequest != null)
                return BadRequest(new { message = "Bạn đã gửi yêu cầu và đang chờ duyệt" });

            var user = await _context.Users.FindAsync(dto.UserId);
            var senderName = user != null ? user.FullName : "Một học sinh";

            var request = new AppRequest
            {
                Title = $"Yêu cầu tham gia {club.Name}",
                Content = string.IsNullOrEmpty(dto.Reason) ? "Xin chào, tôi muốn tham gia câu lạc bộ này." : dto.Reason,
                Category = "CLB_JOIN",
                Status = "Chờ duyệt",
                SenderId = dto.UserId,
                ReceiverIds = club.PresidentId.Value.ToString(),
                ReferenceId = id,
                CreatedDate = DateTime.UtcNow
            };

            _context.AppRequests.Add(request);

            _context.Notifications.Add(new Notification
            {
                UserId = club.PresidentId.Value,
                Title = "Yêu cầu tham gia CLB mới",
                Message = $"{senderName} muốn tham gia {club.Name}.",
                CreatedAt = DateTime.UtcNow
            });

            await _context.SaveChangesAsync();

            return Ok(new { message = "Gửi yêu cầu thành công" });
        }

        // GET: api/club/{id}/pending-requests
        [HttpGet("{id}/pending-requests")]
        public async Task<IActionResult> GetPendingRequests(int id)
        {
            var requests = await _context.AppRequests
                .Include(r => r.Sender)
                .Where(r => r.Category == "CLB_JOIN" && r.ReferenceId == id && r.Status == "Chờ duyệt")
                .Select(r => new
                {
                    r.Id,
                    r.Title,
                    r.Content,
                    r.SenderId,
                    SenderName = r.Sender != null ? r.Sender.FullName : "Unknown",
                    SenderCode = r.Sender != null ? r.Sender.StudentId : "Unknown",
                    Date = r.CreatedDate.ToString("dd/MM/yyyy HH:mm")
                })
                .ToListAsync();

            return Ok(requests);
        }

        // POST: api/club/{id}/transfer-president
        [HttpPost("{id}/transfer-president")]
        public async Task<IActionResult> TransferPresident(int id, [FromBody] TransferPresidentDto dto)
        {
            var club = await _context.Clubs.FindAsync(id);
            if (club == null) return NotFound(new { message = "Không tìm thấy câu lạc bộ" });

            if (club.PresidentId != dto.CurrentPresidentId)
                return BadRequest(new { message = "Bạn không có quyền chuyển chức chủ nhiệm" });

            var newPresident = await _context.Users.FindAsync(dto.NewPresidentId);
            if (newPresident == null)
                return NotFound(new { message = "Không tìm thấy người dùng được chỉ định" });

            club.PresidentId = dto.NewPresidentId;

            // Ensure new president is in members
            if (!await _context.ClubMembers.AnyAsync(cm => cm.ClubId == id && cm.UserId == dto.NewPresidentId))
            {
                _context.ClubMembers.Add(new ClubMember
                {
                    ClubId = id,
                    UserId = dto.NewPresidentId,
                    JoinedDate = DateTime.UtcNow
                });
            }

            _context.Notifications.Add(new Notification
            {
                UserId = dto.NewPresidentId,
                Title = "Chuyển giao quyền Chủ nhiệm",
                Message = $"Bạn đã được bổ nhiệm làm Chủ nhiệm {club.Name}.",
                CreatedAt = DateTime.UtcNow
            });

            await _context.SaveChangesAsync();
            return Ok(new { message = "Chuyển quyền thành công" });
        }
    }

    public class ClubCreateDto
    {
        public string Name { get; set; } = string.Empty;
        public string Type { get; set; } = string.Empty;
        public string Status { get; set; } = "HOẠT ĐỘNG";
        public string Description { get; set; } = string.Empty;
        public int? PresidentId { get; set; }
    }

    public class ClubUpdateDto
    {
        public string Name { get; set; } = string.Empty;
        public string Type { get; set; } = string.Empty;
        public string Status { get; set; } = string.Empty;
        public string Description { get; set; } = string.Empty;
        public int? PresidentId { get; set; }
    }

    public class AddMemberDto
    {
        public int UserId { get; set; }
    }

    public class JoinRequestDto
    {
        public int UserId { get; set; }
        public string Reason { get; set; } = string.Empty;
    }

    public class TransferPresidentDto
    {
        public int CurrentPresidentId { get; set; }
        public int NewPresidentId { get; set; }
    }
}
